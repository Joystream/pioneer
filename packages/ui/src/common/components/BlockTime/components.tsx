import styled from 'styled-components'

import { LabelLink } from '@/common/components/forms'
import { TextMedium, TextSmall } from '@/common/components/typography'
import { Colors, Transitions } from '@/common/constants'

export const AboutText = styled(TextMedium)`
  color: ${Colors.Black[600]};
`
export const BlockInfo = styled.span`
  display: grid;
  grid-auto-flow: column;
  grid-column-gap: 4px;
  align-items: center;
  width: fit-content;
  height: fit-content;
  color: ${Colors.Black[400]};
`
export const BlockNetworkInfo = styled(TextSmall)`
  color: ${Colors.Black[400]};
`
export const BlockNumber = styled(LabelLink)`
  font-size: inherit;
  line-height: inherit;
  transition: ${Transitions.all};
`
