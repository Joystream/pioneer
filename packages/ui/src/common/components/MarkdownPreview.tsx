import React from 'react'
import ReactMarkdown from 'react-markdown'

export const MarkdownPreview = ({ markdown }: { markdown: string }) => <ReactMarkdown>{markdown}</ReactMarkdown>
