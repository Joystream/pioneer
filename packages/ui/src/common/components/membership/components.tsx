import styled, { css } from 'styled-components'

import { BorderRad, Colors, Fonts, Transitions } from '../../../app/constants'
import { AvatarImg } from '../Avatar'
import { Help } from '../Help'
import { TextSmall } from '../typography'
import { MemberInfoWrapProps } from './types'

export const MemberHandle = styled.span`
  grid-area: memberhandle;
  max-width: 100%;
  font-size: 14px;
  line-height: 20px;
  font-weight: 700;
  font-family: ${Fonts.Grotesk};
  color: ${Colors.Black[900]};
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  cursor: pointer;
  transition: ${Transitions.all};

  &:hover {
    color: ${Colors.Black[700]};
  }
`

export const MemberIcons = styled.div`
  display: grid;
  grid-area: membericons;
  grid-auto-flow: column;
  grid-column-gap: 4px;
  align-items: center;
  width: fit-content;
  height: fit-content;
  margin-left: -4px;
`

export const MemberRoles = styled.div`
  display: grid;
  grid-area: memberroles;
  grid-auto-flow: column;
  grid-column-gap: 4px;
  align-items: center;
  width: fit-content;
`

export const MemberId = styled(TextSmall)`
  grid-area: memberroles;
  color: ${Colors.Black[400]};
`

export const MemberPhoto = styled.div`
  display: flex;
  position: relative;
  justify-content: center;
  align-items: center;
  width: 40px;
  height: 40px;
  grid-area: memberphoto;
  border-radius: ${BorderRad.round};
  overflow: hidden;
  border: 1px solid ${Colors.Black[200]};

  ${AvatarImg} {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
`

export const MemberStatusHelp = styled(Help)`
  background-color: transparent;
  color: ${Colors.Blue[500]};

  &:hover,
  &:focus {
    background-color: ${Colors.Blue[500]};
    border-color: ${Colors.Blue[500]};
    color: ${Colors.Black[25]};
  }
`

export const MemberRoleHelp = styled(Help)`
  background-color: ${Colors.Black[100]};
  color: ${Colors.Black[600]};
  border-color: ${Colors.Black[100]};

  &:hover,
  &:focus {
    background-color: ${Colors.Blue[500]};
    border-color: ${Colors.Blue[500]};
    color: ${Colors.Black[25]};
  }
`

export const MemberInfoWrap = styled.div<MemberInfoWrapProps>`
  display: grid;

  grid-template-areas:
    'memberphoto memberhandle membericons'
    'memberphoto memberroles memberroles';
  align-items: center;
  width: 100%;
  justify-self: start;
  border-radius: ${BorderRad.s};
  transition: ${Transitions.all};

  ${MemberPhoto} {
    width: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '80px'
        case 'm':
        default:
          return '40px'
      }
    }};
    height: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '80px'
        case 'm':
        default:
          return '40px'
      }
    }};
  }

  ${MemberRoles} {
    align-self: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return 'start'
        case 'm':
        default:
          return 'center'
      }
    }};
  }

  ${MemberHandle} {
    font-size: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '20px'
        case 'm':
        default:
          return '14px'
      }
    }};
    align-self: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return 'end'
        case 'm':
        default:
          return 'center'
      }
    }};
  }

  ${MemberIcons} {
    align-self: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return 'end'
        case 'm':
        default:
          return 'center'
      }
    }};
    padding: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '4px 0'
        case 'm':
        default:
          return '0'
      }
    }};
    margin-left: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '-6px'
        case 'm':
        default:
          return '-4px'
      }
    }};
  }

  ${({ isOnDark }) => isOnDark === true && MemberOnDarkStyles}

  ${({ memberSize }) => {
    switch (memberSize) {
      case 'l':
        return MemberLargeElements
      case 'm':
      default:
        return MemberMediumElements
    }
  }}
`

export const MemberDarkHover = css`
  ${MemberInfoWrap} {
    ${MemberHandle} {
      color: ${Colors.Black[50]};
    }
    ${MemberStatusHelp} {
      border-color: ${Colors.Blue[400]};
      color: ${Colors.Blue[400]};
      background-color: transparent;

      &:hover,
      &:focus {
        background-color: ${Colors.Blue[500]};
        border-color: ${Colors.Blue[500]};
        color: ${Colors.Black[300]};
      }
    }
    ${MemberRoleHelp} {
      color: ${Colors.Black[200]};
      background-color: ${Colors.Black[500]};
      border-color: ${Colors.Black[500]};

      &:hover,
      &:focus {
        background-color: ${Colors.Blue[500]};
        border-color: ${Colors.Blue[500]};
        color: ${Colors.Black[300]};
      }
    }
  }
`

const MemberOnDarkStyles = css`
  ${MemberHandle} {
    color: ${Colors.Black[75]};
  }
  ${MemberStatusHelp} {
    color: ${Colors.Blue[500]};
    background-color: transparent;
    border-color: ${Colors.Blue[500]};

    &:hover {
      color: ${Colors.Black[300]};
      background-color: ${Colors.Blue[500]};
      border-color: ${Colors.Blue[500]};
    }
  }
  ${MemberRoleHelp} {
    color: ${Colors.Black[300]};
    background-color: ${Colors.Black[600]};
    border-color: ${Colors.Black[600]};

    &:hover {
      color: ${Colors.Black[300]};
      background-color: ${Colors.Blue[500]};
      border-color: ${Colors.Blue[500]};
    }
  }
`

const MemberLargeElements = css`
  grid-template-columns: 80px auto 1fr;
  grid-template-rows: 36px 38px;
  grid-column-gap: 12px;
  grid-row-gap: 6px;
`

const MemberMediumElements = css`
  grid-template-columns: 40px auto 1fr;
  grid-template-rows: 20px 16px;
  grid-column-gap: 8px;
  grid-row-gap: 4px;
`
