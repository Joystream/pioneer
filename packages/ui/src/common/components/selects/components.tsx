import React, { MouseEventHandler, ReactNode } from 'react'
import styled, { css } from 'styled-components'

import {
  Animations,
  BorderRad,
  Colors,
  Overflow,
  RemoveScrollbar,
  Shadows,
  Sizes,
  Transitions,
  ZIndex,
} from '../../constants'
import { ToggleButton } from '../buttons/Toggle'
import { Arrow, CheckboxIconStyles } from '../icons'
import { TextInlineMedium } from '../typography'

import { OptionProps } from './types'

interface Props {
  children: ReactNode
  onClick: () => void
  disabled?: boolean
  className?: string
}

export const Option = ({ children, onClick, disabled, className }: Props) => (
  <OptionComponentContainer onClick={onClick} disabled={disabled}>
    <OptionComponent disabled={disabled} className={className}>
      {children}
    </OptionComponent>
  </OptionComponentContainer>
)

interface OptionSectionHeaderProps {
  children: ReactNode
}

export const OptionSectionHeader = ({ children }: OptionSectionHeaderProps) => (
  <OptionComponentContainer>
    <OptionHeaderComponent>{children}</OptionHeaderComponent>
  </OptionComponentContainer>
)

interface SelectToggleButtonProps {
  isOpen?: boolean
  disabled?: boolean
  onToggleClick: MouseEventHandler
}

export const SelectToggleButton = ({ isOpen, disabled, onToggleClick }: SelectToggleButtonProps) => (
  <ToggleButton isOpen={isOpen} className="ui-toggle" disabled={disabled} onClick={onToggleClick}>
    <Arrow direction="down" />
  </ToggleButton>
)

export const SelectedOption = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 1fr;
  align-items: center;
  padding: 0 4px 0 16px;
  width: 100%;
`

export const EmptyOption = styled.input`
  font-size: 16px;
  line-height: 24px;
  font-weight: 700;
  color: ${Colors.Black[900]};
  width: 100%;
  height: 100%;
  padding: 16px;
  border: none;
  outline: none;
  background-color: transparent;

  &::placeholder {
    font-size: 14px;
    line-height: 78px;
    font-weight: 400;
    color: ${Colors.Black[400]};
  }

  &:disabled {
    cursor: not-allowed;
  }
`

export const SelectComponent = styled.div`
  display: flex;
  position: relative;
  width: 100%;
  height: 100%;
  align-items: center;

  ${ToggleButton} {
    order: 1;
  }

  &:focus-visible {
    outline: none;
  }

  & > :first-child {
    border-radius: 2px;
  }
`

export const OptionComponentContainer = styled.li<{ disabled?: boolean }>`
  display: flex;
  width: 100%;
  height: 100%;
  border: 1px solid ${Colors.Black[300]};
  border-radius: ${BorderRad.s};
  background-color: transparent;

  ${({ disabled }) => {
    if (disabled) {
      return css`
        pointer-events: none;
        cursor: not-allowed;
      `
    }

    return css`
      &:hover {
        .accountName {
          color: ${Colors.Blue[500]};
        }
      }
    `
  }}
`

export const OptionComponent = styled.div<{ disabled?: boolean }>`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 1fr;
  align-items: center;
  width: 100%;
  height: 100%;
  border: none;
  background-color: ${({ disabled }) => (disabled ? Colors.Black[75] : Colors.White)};
  cursor: pointer;
  border-radius: ${BorderRad.s};
  transition: ${Transitions.all};
  min-height: ${Sizes.selectHeight};
  max-height: ${Sizes.selectHeight};
  padding: 10px 72px 10px 16px;

  &:active,
  &:focus {
    outline: none;
  }
`

export const OptionHeaderComponent = styled(OptionComponent)`
  min-height: auto;
`

export const OptionsListComponent = styled.ul`
  display: grid;
  grid-template-columns: 1fr;
  position: absolute;
  left: 50%;
  top: 100%;
  width: calc(100% + 2px);
  height: auto;
  max-height: calc(${Sizes.selectHeight} * 2.5);
  margin: 0;
  border-radius: ${BorderRad.s};
  border: 1px solid ${Colors.Black[300]};
  border-top: 1px solid ${Colors.Blue[400]};
  background-color: ${Colors.White};
  box-shadow: ${Shadows.select};
  overflow-y: scroll;
  transform: translateX(-50%);
  transition: ${Transitions.all};
  animation: showOptions 0.25s ease;
  cursor: auto;
  z-index: ${ZIndex.contextMenu};
  ${RemoveScrollbar};

  @keyframes showOptions {
    from {
      opacity: 0;
      max-height: 0;
    }
  }

  ${OptionComponentContainer} {
    border-left: none;
    border-right: none;
    border-radius: 0;

    ${OptionComponent} {
      border-radius: 0;
    }
  }

  ${OptionComponentContainer} + ${OptionComponentContainer} {
    margin-top: -1px;
  }
`

export const Selected = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 0 4px 0 16px;
  text-transform: capitalize;
  ${Overflow.FullDots};
  cursor: pointer;
  user-select: none;
`

const OptionFocused = css`
  color: ${Colors.Blue[500]};
  path {
    fill: ${Colors.Blue[500]};
  }
`
export const OptionContainer = styled.div`
  display: grid;
  grid-template-columns: 1fr 16px;
  align-items: center;
  width: 100%;
  padding: 0 16px;
  cursor: pointer;
  text-transform: capitalize;
  transition: ${Transitions.all};
  height: 48px;

  ${({ focus, selected }: OptionProps) => (selected || focus) && OptionFocused}
  &:hover {
    ${OptionFocused}
  }

  ${({ selected }: OptionProps) =>
    selected &&
    css`
      background-color: ${Colors.Blue[50]};
      font-weight: bold;
    `}
  ${CheckboxIconStyles} {
    width: 16px;
    height: 16px;
    ${Animations.showSymbol};
  }

  &,
  ${TextInlineMedium} {
    ${Overflow.FullDots};
  }
`
