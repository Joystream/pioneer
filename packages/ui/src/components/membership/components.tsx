import styled, { css } from 'styled-components'
import { BorderRad, Colors, Fonts, Transitions } from '../../constants'
import { AvatarImg } from '../Avatar'
import { Text } from '../typography'
import { MemberInfoWrapProps } from './types'

export const MemberHandle = styled.span`
  grid-area: memberhandle;
  max-width: 100%;
  font-size: 14px;
  line-height: 20px;
  font-weight: 700;
  font-family: ${Fonts.Grotesk};
  color: ${Colors.Black[900]};
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  cursor: pointer;
  transition: ${Transitions.all};

  &:hover {
    color: ${Colors.Black[700]};
  }
`

export const MemberIcons = styled.div`
  display: grid;
  grid-area: membericons;
  grid-auto-flow: column;
  grid-column-gap: 4px;
  align-items: center;
  width: fit-content;
  height: fit-content;
  margin-left: -4px;
`

export const MemberRoles = styled.div`
  display: grid;
  grid-area: memberroles;
  grid-auto-flow: column;
  grid-column-gap: 4px;
  align-items: center;
  width: fit-content;
`

export const MemberRole = styled.span`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 16px;
  height: 16px;
  border-radius: ${BorderRad.round};
  background-color: ${Colors.Black[100]};
  font-size: 6px;
  line-height: 1;
  font-family: ${Fonts.Inter};
  font-weight: 700;
  color: ${Colors.Black[600]};
  text-transform: uppercase;
`

export const MemberId = styled(Text)`
  grid-area: memberroles;
  color: ${Colors.Black[400]};
`

export const MemberPhoto = styled.div`
  display: flex;
  position: relative;
  justify-content: center;
  align-items: center;
  width: 40px;
  height: 40px;
  grid-area: memberphoto;

  ${AvatarImg} {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
`

const MemberOnDarkStyles = css`
  ${MemberHandle} {
    color: ${Colors.Black[75]};
  }
  ${MemberIcons} {
    .memberCircle {
      stroke: ${Colors.Blue[500]};
      transition: ${Transitions.all};
    }
    .memberInner {
      fill: ${Colors.Blue[500]};
      transition: ${Transitions.all};
    }
  }
  ${MemberRole} {
    color: ${Colors.Black[300]};
    background-color: ${Colors.Black[600]};
    transition: ${Transitions.all};
  }
`

const MemberLargeElements = css`
  grid-template-columns: 80px auto 1fr;
  grid-template-rows: 36px 38px;
  grid-column-gap: 12px;
  grid-row-gap: 6px;
`

const MemberMediumElements = css`
  grid-template-columns: 40px auto 1fr;
  grid-template-rows: 20px 16px;
  grid-column-gap: 8px;
  grid-row-gap: 4px;
`

export const MemberInfoWrap = styled.div<MemberInfoWrapProps>`
  display: grid;

  grid-template-areas:
    'memberphoto memberhandle membericons'
    'memberphoto memberroles memberroles';
  align-items: center;
  width: 100%;
  justify-self: start;
  border-radius: ${BorderRad.s};
  transition: ${Transitions.all};

  ${MemberPhoto} {
    width: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '80px'
        case 'm':
        default:
          return '40px'
      }
    }};
    height: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '80px'
        case 'm':
        default:
          return '40px'
      }
    }};
  }

  ${MemberRoles} {
    align-self: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return 'start'
        case 'm':
        default:
          return 'center'
      }
    }};
  }

  ${MemberRole} {
    width: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '24px'
        case 'm':
        default:
          return '16px'
      }
    }};
    height: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '24px'
        case 'm':
        default:
          return '16px'
      }
    }};
    font-size: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '10px'
        case 'm':
        default:
          return '6px'
      }
    }};
  }

  ${MemberHandle} {
    font-size: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '20px'
        case 'm':
        default:
          return '14px'
      }
    }};
    align-self: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return 'end'
        case 'm':
        default:
          return 'center'
      }
    }};
  }

  ${MemberIcons} {
    align-self: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return 'end'
        case 'm':
        default:
          return 'center'
      }
    }};
    padding: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '4px 0'
        case 'm':
        default:
          return '0'
      }
    }};
    margin-left: ${({ memberSize }) => {
      switch (memberSize) {
        case 'l':
          return '-6px'
        case 'm':
        default:
          return '-4px'
      }
    }};
  }

  ${({ isOnDark }) => isOnDark === true && MemberOnDarkStyles}

  ${({ memberSize }) => {
    switch (memberSize) {
      case 'l':
        return MemberLargeElements
      case 'm':
      default:
        return MemberMediumElements
    }
  }}
`

export const MemberDarkHover = css`
  ${MemberInfoWrap} {
    ${MemberHandle} {
      color: ${Colors.Black[50]};
    }
    ${MemberIcons} {
      .memberCircle {
        stroke: ${Colors.Blue[400]};
      }
      .memberInner {
        fill: ${Colors.Blue[400]};
      }
    }
    ${MemberRole} {
      color: ${Colors.Black[200]};
      background-color: ${Colors.Black[500]};
    }
  }
`
