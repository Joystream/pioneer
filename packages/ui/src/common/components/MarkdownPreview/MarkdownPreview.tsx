import React, { ReactNode, useEffect, useMemo, useRef, useState } from 'react'
import ReactMarkdown, { Components } from 'react-markdown'
import { Position } from 'react-markdown/lib/ast-to-react'
import { PluggableList } from 'react-markdown/lib/react-markdown'
import { Root } from 'react-markdown/lib/rehype-filter'
import remarkGfm from 'remark-gfm'
import styled from 'styled-components'

import { Mention, MentionType } from '@/common/components/Mention'
import { UserImage } from '@/common/components/UserImage/UserImage'

import { ModeratedItem } from '../ModeratedItem'

import { MarkdownPreviewStyles, MarkdownPreviewStylesProps } from './MarkdownPreviewStyles'

export interface MarkdownPreviewProps extends MarkdownPreviewStylesProps {
  markdown: string
  append?: ReactNode
}

export const MarkdownPreview = ({ markdown, append, ...styleProps }: MarkdownPreviewProps) => {
  const endsWithP = useRef(false)
  const [appendAfter, setAppendAfter] = useState(false)

  useEffect(() => {
    if (append) setAppendAfter(!endsWithP.current)
  }, [markdown, append])

  const rehypePlugins = useMemo((): PluggableList => {
    const checkEndWithP: PluggableList[0] = () => (tree) => {
      const children = (tree as Root).children
      const lastNode = children?.[children.length - 1]
      endsWithP.current = lastNode?.type === 'element' && lastNode?.tagName === 'p'
      return tree
    }
    return append ? [checkEndWithP] : []
  }, [markdown, append])

  const components = useMemo((): Components => {
    const shouldAppend = (sourcePosition?: Position | null): boolean =>
      endsWithP.current && sourcePosition?.end?.offset === markdown.length

    const p: Components['p'] = ({ children, sourcePosition }) => (
      <p>
        {children}
        {shouldAppend(sourcePosition) && <> {append}</>}
      </p>
    )

    return {
      ...(append ? { p } : {}),

      a: ({ children, ...props }) => {
        const href = props.href as string | undefined
        const match = href?.match(/#mention\?(.+)=(.+)$/) ?? []
        const type = mentionTypesMap[match[1]]
        return type && match[2] ? (
          <Mention itemId={match[2]} type={type}>
            {children}
          </Mention>
        ) : (
          <a style={{ overflowWrap: 'break-word' }} href={href} rel="noopener noreferrer" target="_blank">
            {children}
          </a>
        )
      },
      img: (props) => (
        <MarkdownImage
          src={props.src}
          fallbackComponent={<ModeratedItem title="This image was removed by a moderator" />}
        />
      ),
      code: ({ children, inline }) => <code className={inline ? 'inline-code' : 'in-block-code'}>{children}</code>,
    }
  }, [markdown, append])

  const stripBackslashes = (text: string) => text.replace(/\\(.)/gm, '$1')
  return (
    <div className="markdown-preview">
      <MarkdownPreviewStyles {...styleProps} />
      <ReactMarkdown rehypePlugins={rehypePlugins} remarkPlugins={[remarkGfm]} components={components} rawSourcePos>
        {stripBackslashes(markdown)}
      </ReactMarkdown>
      {appendAfter && <p>{append}</p>}
    </div>
  )
}

const MarkdownImage = styled(UserImage)`
  max-width: 100%;
  max-height: 400px;
`

const mentionTypesMap: Record<string, MentionType> = {
  'member-id': 'member',
  'proposal-id': 'proposal',
  'proposal-post-id': 'proposalDiscussionEntry',
  'thread-id': 'forumThread',
  'forum-post-id': 'forumPost',
  'application-id': 'application',
  'opening-id': 'opening',
}
