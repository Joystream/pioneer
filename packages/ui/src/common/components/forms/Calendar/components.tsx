import { addDays, addMonths, isAfter, isBefore, isEqual, isSameMonth } from 'date-fns'
import React, { FC } from 'react'
import styled, { css } from 'styled-components'

import { TextSmall } from '@/common/components/typography'
import { Colors, Fonts } from '@/common/constants'
import { DateSelection, PartialDateRange } from '@/common/types/Dates'
import { fromRange } from '@/common/utils/dates'

import { ButtonGhost, ButtonInnerWrapper } from '../../buttons'

interface CalendarNavProps {
  month: Date
  direction: -1 | 1
  min?: Date
  max?: Date
  onClick?: (month: Date) => void
}

export const CalendarNav: FC<CalendarNavProps> = ({ month, direction, min, max, children, onClick }) => {
  const target = addMonths(month, direction)
  const isOutofBounds = direction > 0 ? max && isAfter(target, max) : min && isBefore(addDays(month, -1), min)
  if (!onClick || isOutofBounds) {
    return <ChangeMonthButton disabled>{children}</ChangeMonthButton>
  } else {
    const changeMonth = () => {
      onClick(addMonths(month, direction))
    }
    return <ChangeMonthButton onClick={changeMonth}>{children}</ChangeMonthButton>
  }
}

interface CalendarDayProps {
  day: Date
  month: Date
  selection?: DateSelection
  within?: PartialDateRange
  onClick?: (day: Date) => void
}

export const CalendarDay = ({ day, month, selection, within, onClick }: CalendarDayProps) => {
  const { start: min, end: max } = fromRange(within)
  const tooEarly = min && isBefore(day, min)
  const tooLate = max && isAfter(day, max)
  if (tooEarly || tooLate) {
    return <DayButton disabled>{day.getDate()}</DayButton>
  }

  return (
    <DayButton
      {...selectionFlags(day, selection)}
      inMonth={isSameMonth(day, month)}
      onClick={onClick && (() => onClick(day))}
    >
      {day.getDate()}
    </DayButton>
  )
}

interface SelectionFlags {
  first?: boolean
  last?: boolean
  selected?: boolean
}

const selectionFlags = (day: Date, selection?: DateSelection): SelectionFlags => {
  if (selection) {
    if (selection instanceof Date) {
      if (isEqual(day, selection)) {
        return { selected: true, first: true, last: true }
      }
    } else if ('start' in selection || 'end' in selection) {
      const after = !('start' in selection) || isAfter(day, selection.start)
      const before = !('end' in selection) || isBefore(day, selection.end)
      const first = 'start' in selection && isEqual(day, selection.start)
      const last = 'end' in selection && isEqual(day, selection.end)
      const selected = (after && before) || first || last
      return { selected, first, last }
    }
  }
  return {}
}

const ITEM_FAKE_SIZE = '33px'
const ITEM_SIZE = '32px'
export const CALENDAR_WRAP_SIZE = '224px'

export const CalendarGrid = styled.div`
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  max-width: ${CALENDAR_WRAP_SIZE};
  user-select: none;
`

export const CalendarMonth = styled(CalendarGrid)`
  justify-content: start;
`

const ItemStyle = css`
  text-align: center;
  vertical-align: middle;
  line-height: ${ITEM_SIZE};
`

export const CalendarHeaderContainer = styled(CalendarGrid)`
  width: 100%;
  background-color: ${Colors.White};
`
export const CalendarTitle = styled(TextSmall)`
  ${ItemStyle}
  width: 160px;
`

export const CalendarWeekDay = styled.span`
  ${ItemStyle}
  display: flex;
  justify-content: center;
  align-items: center;
  width: ${ITEM_FAKE_SIZE};
  height: ${ITEM_FAKE_SIZE};
  text-align: center;
  font-family: ${Fonts.Grotesk};
  font-size: 10px;
  line-height: 16px;
  font-weight: 700;
  color: ${Colors.Black[400]};
  margin-left: -1px;
  margin-top: -1px;
`

const CalendarButton = styled(ButtonGhost).attrs(() => ({
  size: 'small',
  square: true,
}))`
  min-width: ${ITEM_FAKE_SIZE};
  width: ${ITEM_FAKE_SIZE};
  max-width: ${ITEM_FAKE_SIZE};
  min-height: ${ITEM_FAKE_SIZE};
  height: ${ITEM_FAKE_SIZE};
  max-height: ${ITEM_FAKE_SIZE};
  margin-top: -1px;
  margin-left: -1px;
  font-size: 10px;
  line-height: 16px;
  border-radius: 0;

  &:not(:first-child) {
    margin-left: -1px;
  }

  &:focus {
    border-color: ${Colors.Black[200]};
    color: ${Colors.Black[900]};
    &:before {
      transform: translate(-150%, -50%);
    }
  }
`

const ChangeMonthButton = styled(ButtonGhost).attrs(() => ({
  size: 'small',
  square: true,
}))`
  min-width: ${ITEM_SIZE};
  width: ${ITEM_SIZE};
  max-width: ${ITEM_SIZE};
  min-height: ${ITEM_SIZE};
  height: ${ITEM_SIZE};
  max-height: ${ITEM_SIZE};
  font-size: 10px;
  line-height: 16px;
  ${({ disabled }) =>
    disabled &&
    css`
      visibility: hidden;
    `}
`

interface SelectableDayProps extends SelectionFlags {
  inMonth?: boolean
  disabled?: boolean
}

const DayButton = styled(CalendarButton)<SelectableDayProps>`
  font-family: ${Fonts.Inter};
  font-weight: 400;
  width: ${ITEM_FAKE_SIZE};
  height: ${ITEM_FAKE_SIZE};

  ${ButtonInnerWrapper} {
    transform: translateY(0);
  }

  ${({ selected, first, last, inMonth, disabled }: SelectableDayProps) => {
    if (disabled) {
      return css`
        background-color: ${Colors.Black[50]};
        color: ${Colors.Black[200]};
        cursor: not-allowed;
      `
    } else if (!inMonth) {
      return css`
        background-color: ${selected ? Colors.Blue[50] : Colors.Black[50]};
        color: ${Colors.Black[400]};
      `
    } else if (first || last) {
      return css`
        background-color: ${Colors.Blue[500]};
        color: ${Colors.White};
        &:focus {
          color: ${Colors.White};
        }
        ${first &&
        css`
          border-top-left-radius: 50%;
          border-bottom-left-radius: 50%;
        `}
        ${last &&
        css`
          border-top-right-radius: 50%;
          border-bottom-right-radius: 50%;
        `}
      `
    } else if (selected) {
      return css`
        background-color: ${Colors.Blue[50]};
      `
    }
  }}
`
