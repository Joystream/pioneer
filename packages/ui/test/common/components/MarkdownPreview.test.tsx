import { render } from '@testing-library/react'
import React from 'react'

import { MarkdownPreview } from '@/common/components/MarkdownPreview'

const EDITOR_OUTPUT = `## Heading 1

### Heading 2

#### Heading 3

*   ui 1
    *   ul 1.1
        *   ul 1.1.1
    *   ul 1.2
*   ul 2

1.  ol 1
2.  ol 2

> A quote`

describe.skip('UI: MarkdownPreview', () => {
  it('Empty', async () => {
    const { container } = renderComponent('')

    expect(container.innerHTML).toBe('')
  })

  it('Single paragraph', async () => {
    const { container } = renderComponent('Foo bar baz')

    expect(container.innerHTML).toBe('<p>Foo bar baz</p>')
  })

  it('Example editor output', async () => {
    const { container } = renderComponent(EDITOR_OUTPUT)

    expect(container.innerHTML).toMatchSnapshot()
  })

  function renderComponent(markdown: string) {
    return render(<MarkdownPreview markdown={markdown} />)
  }
})
