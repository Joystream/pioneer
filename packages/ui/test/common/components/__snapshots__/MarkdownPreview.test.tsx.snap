// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI: MarkdownPreview Example editor output 1`] = `
"<h2>Heading 1</h2>
<h3>Heading 2</h3>
<h4>Heading 3</h4>
<ul>
<li>ui 1
<ul>
<li>ul 1.1
<ul>
<li>ul 1.1.1</li>
</ul>
</li>
<li>ul 1.2</li>
</ul>
</li>
<li>ul 2</li>
</ul>
<ol>
<li>ol 1</li>
<li>ol 2</li>
</ol>
<blockquote>
<p>A quote</p>
</blockquote>"
`;
